lib/Dialect/Seq/SeqOps.cpp

```diff
@@ -658,6 +661,120 @@ LogicalResult ClockGateOp::canonicalize(ClockGateOp op,
+}
+
+template <class Op>
+static LogicalResult verifyFirMemMask(Op op) {
+  if (auto mask = op.getMask()) {
+    auto memType = op.getMemory().getType();
+    if (!memType.getMaskWidth())
+      return op.emitOpError("has mask operand but memory type '")
+             << memType << "' has no mask";
+    auto expected = IntegerType::get(op.getContext(), *memType.getMaskWidth());
+    if (mask.getType() != expected)
+      return op.emitOpError("has mask operand of type '")
+             << mask.getType() << "', but memory type requires '" << expected
+             << "'";
+  }
+  return success();
+}
+
+LogicalResult FirMemWriteOp::verify() { return verifyFirMemMask(*this); }
+LogicalResult FirMemReadWriteOp::verify() { return verifyFirMemMask(*this); }
+
+static bool isConst(Value value) {
+  if (value)
+    return value.getDefiningOp<hw::ConstantOp>();
+  return false;
+}
+
+static bool isConstZero(Value value) {
+  if (value)
+    if (auto constOp = value.getDefiningOp<hw::ConstantOp>())
+      return constOp.getValue().isZero();
+  return false;
+}
+
+static bool isConstAllOnes(Value value) {
+  if (value)
+    if (auto constOp = value.getDefiningOp<hw::ConstantOp>())
+      return constOp.getValue().isAllOnes();
+  return false;
+}
+
+LogicalResult FirMemReadOp::canonicalize(FirMemReadOp op,
+                                         PatternRewriter &rewriter) {
+  // Remove the enable if it is constant true.
+  if (isConstAllOnes(op.getEnable())) {
+    rewriter.updateRootInPlace(op, [&] { op.getEnableMutable().erase(0); });
+    return success();
+  }
+  return failure();
+}
+
+LogicalResult FirMemWriteOp::canonicalize(FirMemWriteOp op,
+                                          PatternRewriter &rewriter) {
+  // Remove the write port if it is trivially dead.
+  if (isConstZero(op.getEnable()) || isConstZero(op.getMask()) ||
+      isConst(op.getClock())) {
+    rewriter.eraseOp(op);
+    return success();
+  }
+  bool anyChanges = false;
+
+  // Remove the enable if it is constant true.
+  if (auto enable = op.getEnable(); isConstAllOnes(enable)) {
+    rewriter.updateRootInPlace(op, [&] { op.getEnableMutable().erase(0); });
+    anyChanges = true;
+  }
+
+  // Remove the mask if it is all ones.
+  if (auto mask = op.getMask(); isConstAllOnes(mask)) {
+    rewriter.updateRootInPlace(op, [&] { op.getMaskMutable().erase(0); });
+    anyChanges = true;
+  }
+
+  return success(anyChanges);
+}
+
+LogicalResult FirMemReadWriteOp::canonicalize(FirMemReadWriteOp op,
+                                              PatternRewriter &rewriter) {
+  // Replace the read-write port with a read port if the write behavior is
+  // trivially disabled.
+  if (isConstZero(op.getEnable()) || isConstZero(op.getMask()) ||
+      isConst(op.getClock()) || isConstZero(op.getMode())) {
+    auto newOp = rewriter.replaceOpWithNewOp<FirMemReadOp>(
+        op, op.getMemory(), op.getAddress(), op.getClock(), op.getEnable());
+    for (auto namedAttr : op->getAttrs())
```

Comment:
I'm getting a crash here when running the canonicalization tests locally. Shouldn't you not be accessing `op` after replacing it (not sure why it passes in CI)?

Reply:
Opened pr to address this [here](https://github.com/llvm/circt/pull/5460)